rfcs/enforce-legibility.md

```diff
@@ -0,0 +1,247 @@
+         ; ... it only exports direct symbols, not from submodules, and...
+         (not exports-subvalues)
+         ; ... it does not exports types (avoid using ppx_import).
+         (not exports-types)))
+
+  ; - removed and replaced by local 'let open <X> in's if...
+  ;   ... it is used only by only a few functions.
+  (local (<= functions 4))
+
+  ; - moved closer to its first actual use if...
+  ;   ... it is too far from that optimal placement.
+  (move (>= dist-to-optimal 40)))
+```
+
+This makes the rules easily modifiable in the future, along with the evolution of the tool and the consensus of the team, and we propose that future changes of rules can be made with simple pull requests rather than new RFCs. The current rules can always be gathered by consulting the `.ocamlclose` file.
+
+### Integration into the development pipeline
+
+We propose that the tool is run with the aforementioned rules **as a pre-commit hook**, only on files that are about to be committed. This allows for an incremental removal of problematic `open`s instead of blocking development (with the rules exposed above, several hundreds of problematic `open`s are detected in the whole code base).
+
+The tool will block the commit if problems are found, and suggest modifications.
+It is then up to the developer to automatically apply the suggested patches, manually fix the problem or override the tool's decision by white-listing the modules.
+
+White-listing a module can be done by either modifying the global rule or dropping a new `.ocamlclose` encoding the white-listing. The module will thus only be white-listed in the directory and subdirectories of the new file, without affecting the global rules. Since it is encoded in the file-system, the white-listing can then be debated during a code review. 
+
+## Examples
+
+The current above configuration contains rules for **removing** (`remove`), **narrowing** (`structure`), **localizing** (`local`) or **moving** (`move`) open statements. They are applied by running the following commands:
+
+```bash
+ocamlclose lint -p fixes file1.ml file2.ml # Prints a set of recommendations, saved in a 'fixes' file
+ocamlclose patch -ic fixes                 # Modifies the original file in place, checking they can still build
+```
+
+Here are some use cases for each rule.
+
+### Remove
+
+Some open statements are just not used enough to justify their existence. For example, the file `src/lib/work_selector/work_lib.ml` contains an `open Currency` which is just used once. `ocaml-close` yields the following patch:
+
+```diff
+ open Core_kernel
+-open Currency
+ open Async
+ 
+ module Make (Inputs : Intf.Inputs_intf) = struct
+@@ -123,7 +122,7 @@
+       (Inputs.Snark_pool.get_completed_work snark_pool statements)
+       ~f:(fun priced_proof ->
+         let competing_fee = Inputs.Transaction_snark_work.fee priced_proof in
+-        Fee.compare fee competing_fee < 0)
++        Currency.Fee.compare fee competing_fee < 0)
```

Comment:
`Currency` is one of our 'standard' `open`s, and really should be open everywhere, for consistency. That is, `Fee`, `Balance`, and `Amount` should always resolve to `Currency.Fee` etc. (except where a functor needs special treatment).

Can you add `Currency` to the exclusion list, and update with a desirable example?